### DISCUSSION TITLE
Add markdown file full support

### DISCUSSION THREAD

**Discussion by PavelTurk** [apache/netbeans#3477] [Ideas]
### Description

I use Apache NetBeans IDE 12.5 and highlighting doesn't work for markdown files. Besides, when readme.md is in the root of the project (where it usually is) NetBeans doesn't show the file on Projects tab - it is necessary to open Files tab. 

And of course, when we talk about markdown support it means that it must have preview. Without preview it doesn't have any sense. So, I suggest to add a full support for markdown files.

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow the Apache Software Foundation's [Code of Conduct](https://www.apache.org/foundation/policies/conduct.html)

**Comment by geertjanw** [comment]
I believe this should do it, please come back to this discussion and share your findings, or fork this and develop it further: 

https://github.com/madflow/flow-netbeans-markdown

**Reply by PavelTurk** [reply]
@geertjanw Thank you for the quick answer.

It is possible to use the plugin you linked, however, there are some problems with it:

1.  If NB is running on Java 14, it is not possible to install it. It gives:
```
The validation of downloaded plugins cannot be completed, cause: NBM /home/pavel/.netbeans/12.5/update/download/flow-netbeans-markdown.nbm needs unpack200 to process following entries: 
netbeans/modules/ext/asm-4.1.jar.pack.gz
netbeans/modules/ext/asm-analysis-4.1.jar.pack.gz
netbeans/modules/ext/asm-tree-4.1.jar.pack.gz
netbeans/modules/ext/asm-util-4.1.jar.pack.gz
netbeans/modules/ext/parboiled-core-1.1.7.jar.pack.gz
netbeans/modules/ext/parboiled-java-1.1.7.jar.pack.gz
netbeans/modules/ext/pegdown-1.5.0.jar.pack.gz
netbeans/modules/ext/pegdown-1.6.0.jar.pack.gz
netbeans/modules/flow-netbeans-markdown.jar.pack.gz
```
so, it is necessary to move to Java 11, install plugin and move back to Java 14

2. md files located in the project root, for example, readme.md are not displayed on Projects tab.
3. Plugin project doesn't seem to be supported anymore - last commit 2018 year.
4. This plugin has a serious bug - when you click `Preview` window is scrolled to the end, see my issue [here](https://github.com/madflow/flow-netbeans-markdown/issues/131)
5. After installing the plugin there are two source buttons:
![Screenshot from 2022-01-19 14-17-59](https://user-images.githubusercontent.com/10261361/150128041-4ffbe406-4878-4424-9957-f8f0f2547af6.png)

Generally, this plugin is not bad, it works, but it is necessary to fix all its bugs (I just don't remember all of them now).

**Comment by geertjanw** [comment]
Great, you're welcome to contact the plugin owner, and work with them on this, and/or ask whether the sources can be contributed to Apache NetBeans so that we can work together on them here.

**Reply by PavelTurk** [reply]
When I opened the issue I got no answer from them, and not only I. Maybe, if you or any other from Apache Software Foundation ask them to contribute their sources they will agree. At least, there is a higher probability that they will answer.

**Reply by geertjanw** [reply]
Done.

**Reply by madflow** [reply]
Like stated here: https://github.com/madflow/flow-netbeans-markdown/issues/128 - we (the main contributors) are open to "donate" the source code. @PashaTurok  not sure where/who you asked, but the linked issue is pretty clear on that I would argue. I do not use Netbeans anymore, so I have no interest in developing it anymore. Forks are welcome.

**Comment by matthiasblaesing** [comment]
For the record: Markdown "syntax highlighting" is here: https://github.com/apache/netbeans/tree/master/ide/markdown. The big question would be what you expect from syntax highlighting and it might mean, that you have to redo the whole work again with a different approach. What is currently missing is a preview. You can use a java library to render the markdown to HTML and use WebView or even Swing TextView to render the result.

**Reply by PavelTurk** [reply]
No, if NetBeans is an IDE, then it must have an internal markdown file preview. We mustn't use any external programs or other IDE's to preview a markdown files. What about "syntax highlighting" I didn't see any "syntax highlighting" in NB 12.5. As I understand NB must support highlighting from 12.4, I am repeating I didn't see it in 12.5. At the same time, maybe we are talking about different things. 

For example, this is a "syntax highlighting" of flow-netbeans-markdown plugin for markdown file [demo](https://markdown-it.github.io/)

![Screenshot from 2022-01-20 22-06-33](https://user-images.githubusercontent.com/10261361/150414159-fb103fb5-e164-42da-8e89-049cd240bbf8.png)

**Reply by matthiasblaesing** [reply]
Please read again what I wrote. I never said something about external programms. NetBeans is shipped with the JavaFX based webview, that creates a HTML renderer, that works inside the Window and can be embedded in Swing. 

It is useless to write another Markdown Renderer if there are a several good libraries to choose from. 

There is highlighting, just not what you expect (the screenshot is from current master):

![image](https://user-images.githubusercontent.com/2179736/150417410-e795c435-56a9-4938-8efa-a6ab6299b7ef.png)

**Reply by PavelTurk** [reply]
I am not a NB developer, could you just explain where I should click to preview my markdown file and in what NB version? I didn't see the highlighting you provided in NB 12.5

**Reply by matthiasblaesing** [reply]
There is syntax highlighting. You are seeing the result of me creating the file "test.md" and copying the sample markdown content into it. End of story, nothing to be done here. A preview is not there yet, it can be implemented.

**Reply by PavelTurk** [reply]
I don't agree that this is the end of the story. I don't understand "preview can be implemented". It must be implemented otherwise it is a not a full support for markdown files. The issue is about a full support.

**Reply by neilcsmith-net** [reply]
> NetBeans is shipped with the JavaFX based webview, that creates a HTML renderer, that works inside the Window and can be embedded in Swing.

To be pedantic, NetBeans ships with the ability to download JavaFX webview on demand.  We can never ship with it because it's LGPL.  And that download model causes enough issues already.  So, any Markdown preview using this would have to be carefully designed to be optional.

Personally I'm for supporting live preview in the browser, like other markup, but then that's the workflow I'm already used to.  Most of my own use of Markdown in NetBeans is with Jekyll or MkDocs, which support this well already, and where any preview would be missing a lot.

**Comment by PavelTurk** [comment]
@geertjanw I suggest to reopen the issue, as there is nor preview nor syntax highlighting at least in NB 12.5.

**Comment by Chris2011** [comment]
Hey @PashaTurok thx for your ticket and the discussion. I can understand your frustration, so let me just try to make some things clearer.

So with NetBeans 12.* somehow, markdown syntax highlighting was added but just basic one via textmare lexer files like in VS Code or others who uses textmate nowadays. So this is very very basic and not super highlighted what you and me would expect.

To make it clear, I have the same file, side by side left is the core NetBeans support and right is the flow-markdown plugin support:
![image](https://user-images.githubusercontent.com/795658/151395954-52638b3c-051c-42b1-b504-ecd4a9ecdecf.png)


So yeah there are differences. But again this was implemented. So I would say this will be another ticket for "making the highlighting better".

What you mean by full support could be a lot and we need to clarify it, not just in one ticket.
I can tell you, what I think could be "full".

* Powerfull syntax highlighting (Maybe semantic too
* Showing a preview (side by side, splitview)
* Rendering while typing (and not jump to the end/beginning as the flow-markdown plugin does)
* Synchronize scrolling (If you scroll in source, the preview should also scroll and vice versa)
* Support Markdown editing right inside the preview (That means a new editor somehoe like this cool thing: https://www.typora.io/) So somehow formatting after writing the correct markdown stuff.
* Code completion for HTML (And maybe other languages, it matters what Markdown can handle)
* Use another, more common, more maintained markdown parser
* to be continued

So as you can see, there is a lot what we would like to have in a full or more powerfull markdown support in NetBeans. Unfortunately, priorities are different and me and also another one tried to fix one or two things but not for the core, more or less for the plugin. A lot of stuff were already created as issues in the flow-markdown plugin as you can see.

At the end, all those stuff is nice but shouldn't be handled in just one ticket like "Add full markdown support". Not clear enough what you meant by full. So for those stuff, the discussion section is here or better to make a proposal: https://cwiki.apache.org/confluence/display/NETBEANS/Feature+Request+Outlines what you mean and what could be a full markdown support. Hope that makes things a bit clearer.

**Reply by neilcsmith-net** [reply]
Thanks @Chris2011 that's a good summary.  One other bullet point I meant to look at myself is spell checking in Textmate support.  I quite often edit Markdown in NetBeans myself.  I sometimes uninstall the support we do have to get spell checking back.

**Reply by Chris2011** [reply]
Yes, spellchecking is also a nice a very nice thing.
